src/v/coproc/reconciliation_backend.cc

```diff
@@ -36,94 +62,112 @@ reconciliation_backend::reconciliation_backend(
       [this](std::vector<update_t> deltas) {
-          return within_context([this, deltas = std::move(deltas)]() mutable {
-              for (auto& d : deltas) {
-                  auto ntp = d.ntp;
-                  _topic_deltas[ntp].push_back(std::move(d));
-              }
-              return fetch_and_reconcile();
-          });
+          if (!_gate.is_closed()) {
+              enqueue_events(std::move(deltas));
+          } else {
+              vlog(
+                coproclog.debug,
+                "Gate closed encountered while handling delta notification");
+          }
       });
+    try {
+        (void)ss::with_gate(_gate, [this] {
+            return ss::do_until(
+              [this] { return _as.abort_requested(); },
+              [this] { return fetch_and_reconcile(std::move(_topic_deltas)); });
+        });
+    } catch (const ss::gate_closed_exception&) {
+        vlog(
+          coproclog.debug,
+          "Gate closed encountered when attempting to start reconciliation "
+          "backend");
+    }
     return ss::now();
 }
 
 ss::future<> reconciliation_backend::stop() {
-    _retry_timer.cancel();
     _as.request_abort();
     _topics.local().unregister_delta_notification(_id_cb);
     return _gate.close();
 }
 
-ss::future<std::vector<reconciliation_backend::update_t>>
-reconciliation_backend::process_events_for_ntp(
+ss::future<> reconciliation_backend::process_updates(
   model::ntp ntp, std::vector<update_t> updates) {
-    std::vector<update_t> retries;
-    for (auto& d : updates) {
-        vlog(coproclog.trace, "executing ntp: {} op: {}", d.ntp, d);
-        auto err = co_await process_update(d);
-        vlog(coproclog.info, "partition operation {} result {}", d, err);
-        if (err == errc::partition_not_exists) {
-            /// In this case the source topic exists but its
-            /// associated partition doesn't, so try again
-            retries.push_back(std::move(d));
+    auto itr = updates.begin();
+    while (itr != updates.end()) {
+        vlog(coproclog.trace, "executing ntp: {} op: {}", ntp, *itr);
+        auto err = co_await process_update(ntp, *itr);
+        vlog(coproclog.info, "partition operation {} result {}", *itr, err);
+        /// This retry mechanism exists to replay events that must be
+        /// executed before progress can be made.
+        ///
+        /// An example is creating a materialized partition. This action
+        /// depends on the cluster::controller_backend having first
+        /// created the source partition. This is an event that MUST have
+        /// occurred first in absolute time otherwise an event to create a
+        /// materialized log with that source as a parent would never have
+        /// been fired.
+        if (err != errc::partition_not_exists) {
+            ++itr;
+        } else {
+            vlog(
+              coproclog.warn,
+              "There were recoverable errors when processing events, "
```

Comment:
there is only one error causing this update to be retried, maybe we should add this information to the log entry ?